packages/aws-cdk-lib/aws-ecr/lib/repository.ts

```diff
@@ -224,10 +231,31 @@ export abstract class RepositoryBase extends Resource implements IRepository {
    * @param tagOrDigest Optional image tag or digest (digests must start with `sha256:`)
    */
   public repositoryUriForTagOrDigest(tagOrDigest?: string): string {
-    if (tagOrDigest?.startsWith('sha256:')) {
-      return this.repositoryUriForDigest(tagOrDigest);
+    if (tagOrDigest && Token.isUnresolved(tagOrDigest)) {
+
+      let isInputDigestCondition;
+
+      // Use the existing condition of the same Token is existed, and if not create a new one
+      if (tagOrDigest in this.tokenConditions) {
+        isInputDigestCondition = this.tokenConditions[tagOrDigest];
+      } else {
+        this.conditionsNumber += 1;
+        isInputDigestCondition = new CfnCondition(this, `IsInputDigest${this.conditionsNumber}`, {
+          // we split the value of the Token using the delimiter : to check if the first element equals sha256
+          // to check if it is a digest, or it will be an image tag.
+          expression: Fn.conditionEquals(Fn.select(0, Fn.split(':', tagOrDigest)), 'sha256'),
+        });
+        this.tokenConditions[tagOrDigest] = isInputDigestCondition;
+      }
+
+      const suffix = Fn.conditionIf(isInputDigestCondition.logicalId, `@${tagOrDigest}`, `:${tagOrDigest}`).toString();
```

Comment:
What is this line doing? What's the purpose of `@${tagOrDigest}` and `:${tagOrDigest}` being passed in?

Reply:
the condition I added checks if the passed Token contains a Tag or Image Digest, if it is a tag the repo uri should be in this format `<ecr repo base uri>@<image digest value>`, if it is a tag, so the format will be `<ecr repo base uri>:<tag value>`

Reply:
I tagged you in 2 examples for each use case in the integ testing template

Reply:
I can update the code as well to make it more clear